dtype fixes for downsample and normalization

ivirshup opened the pull request:
Addressing https://github.com/theislab/scanpy/issues/435#issuecomment-538776417

This PR does two things:

1. `downsample_counts` will convert the resulting downsampled matrix back to the initial dtype by default.
2. `normalize_total` will now work with integer matrices.

I think 2 should definitely be the case. 1 does have a performance cost, but it's close to @falexwolf's [suggestion](https://github.com/theislab/scanpy/issues/435#issuecomment-475999342) and removes a minor foot-gun.

ivirshup committed:
Have downsample return the same dtype it was given by default

ivirshup committed:
Make normalize total work with integer X

ivirshup committed:
Fix normalize total for layers arugment, and test it

* Also reformat

ivirshup commented:
I'm thinking about reverting the return dtype part, but just keeping having `normalize_total` to work with more dtypes. Thoughts @flying-sheep?

flying-sheep commented on scanpy/preprocessing/_normalization.py:
```diff
@@ -11,14 +11,16 @@
     else:
-        X /= counts[:, None]
-    return X if copy else None
+        np.divide(X, counts[:, None], out=X)
```
Why is this and the above change necessary? I though that if `__itruediv__` is not available, python falls back to `__truediv__` anyway?

ivirshup replied:
Sorry for the late response on this, totally forgot about it.

The conversion of `X` to float happens because you can't do an inplace dtype conversion. For example:

```
>>> a = np.ones(3, dtype=int)
>>> np.divide(a, 2, out=a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: No loop matching the specified signature and casting was found for ufunc true_divide
```

This is also the reason behind changing `count /= after` with `counts = counts / after`. If `counts` was generated by `X.sum`, it could be an integer array. Which would do:

```
>>> a /= 2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: No loop matching the specified signature and casting was found for ufunc true_divide
```

Finally, I replaced `X /= counts[:, None]` with `np.divide(X, counts[:, None], out=X)` because if X was dask array `__itruediv__` doesn't behave like it does for numpy arrays https://github.com/dask/dask/issues/5199.

flying-sheep replied:
Thank you! Would be cool to have this documented in a comment

flying-sheep reviewed (commented):
I need more info: why did you do the changes you made in `_normalize_data` and `normalize_total`?

And what are the arguments for/against maintaining the dtype?

ivirshup commented:
Arguments for and against converting values in `downsample_counts`:

If we don't convert dtypes back to what they originally were, there's a slight performance boost since we don't have to have two copies.

I we return an array of integers we run into trouble downstream with functions that aren't tested with integer arrays. Issues from this have been opened a few times, so when I wrote this I thought it might be worth just maintaining the input type. I'm not sure I agree with that now.

LuckyMD commented:
> If we return an array of integers we run into trouble downstream with functions that aren't tested with integer arrays. Issues from this have been opened a few times, so when I wrote this I thought it might be worth just maintaining the input type. I'm not sure I agree with that now.

This is quite a compelling argument for me (as I was one of the people who reported an issue like this). If an integer matrix is generally returned, then one would have to ensure all other functions will work with this data type as intended (sc.pp.log1p for example). Otherwise this would be backward-breaking.

GMaciag commented:
I would just like to add that the issue with `normalize_total` can arise not only from the `downsample_counts` function. 

In my case, I am working on `.loom` files generated with *velocyto* - I want to be able to estimate RNA velocity in the end. That means I have 'spliced' and 'unspliced' layers in my anndata object. I wanted to use `normalize_total` on all the layers, which should be possible by setting parameter `layers='all'`. However, I was getting a TypeError, as in  #435 

The workaround described at the end of that issue solved it for me. My point is just that fixing only `downsample_counts` is not enough and functions that work on layers should accept integer data. I think that my case is not that uncommon and will happen more often as people use scVelo with velocyto pipelines. 

Alternatively, you should warn people about it in the tutorials and make them convert everything to float.

ivirshup committed:
Minor bugfix

* Test wasn't creating anndata object with array of correct type
* `normalize_total` wasn't playing nice with dask since dask doesn't do the single argument version of `np.median`.

flying-sheep commented:
We should definitely maintain the type in layers, and that means maintaining the type in .X makes sense too.

We should also take care not to downcast more incompatible types: int32 can be expressed as float64, but not in float32. int64 has to stay int64.

ivirshup committed:
Improve handling of floats for downsample_counts

Now the internal code can just handle floats directly, so we don't have to worry about type conversion. Tests have been added to account for this as well.

ivirshup committed:
Formatting pass on downsample_counts.

ivirshup commented:
Hah, I've gotten much better at numba since I wrote this function. I figured out I can just get the core part to work on floats and don't have to worry about casting between types. Makes this a much easier decision. Now floats aren't converted to integers in the first place.

> We should also take care not to downcast more incompatible types: int32 can be expressed as float64, but not in float32. int64 has to stay int64.

I think we can be a little flexible on this, and just generally follow numpy promotion rules (except for when they're bad).

flying-sheep commented:
[these ones](https://docs.scipy.org/doc/numpy/reference/generated/numpy.result_type.html). But you still convert ints to floats? And when are they bad?

ivirshup commented:
> But you still convert ints to floats?

`downsample_counts` now doesn't convert anything, or even copy if you don't want it to. It just modifies the passed matrix (or copy, if `copy=True`) whether it has integer or floating point values.

> And when are they bad?

This one is pretty weird:

```python
np.result_type(np.int64, np.uint64)
# dtype('float64')
```

But also numpy does promotion (or some other conversion, not quite sure for this one) when it probably shouldn't:

```python
type(np.uint8(255) << 1)
# numpy.int64
```

flying-sheep commented:
> This one is pretty weird:
> 
> ```py
> np.result_type(np.int64, np.uint64)
> # dtype('float64')
> ```

That’s also a bug, the following code changes the first integer (+=1) and completely destroys the second one, without need.

```py
>>> infos = [np.iinfo(dt).max for dt in [np.int64, np.uint64]]
>>> infos
[9223372036854775807, 18446744073709551615]
>>> np.array(infos)
array([9.22337204e+18, 1.84467441e+19]
>>> np.array(infos).astype(np.uint64)
array([9223372036854775808, 0], dtype=uint64)
```

flying-sheep commented on scanpy/preprocessing/_normalization.py:
```diff
@@ -11,14 +11,17 @@
     X = X.copy() if copy else X
-    after = np.median(counts[counts>0]) if after is None else after
+    if issubclass(X.dtype.type, (int, np.integer)):
+        X = X.astype(np.float32)  # TODO: Check if float64 should be used
```
yeah, we should at least implement that TODO, leaking precision shouldn’t be done willy-nilly.

Unless that’s a huge performance cost of course.

ivirshup replied:
Oh, by check I meant do a benchmarking and seeing what range of values would be required for imprecision. Looking at the functions we're calling, it's all numpy and nothing is iterative on our end, so I'm not too worried about it. I think it'll be no worse than our default behavior.

What were you thinking for implementation? int64 -> float64 or something like that?

flying-sheep replied:
Everything is simplified by the fact that 1) scRNASeq data is pretty robust to rounding [citation needed] and 2) neither counts nor any normalization we do go far up. it still feels weird:

- float32 can represent ints <= 16777216 precisely
- int64 can represent ints <= 9223372036854775807

I’d sleep more easily if I knew what order of magnitude the highest number has that exists at any point in scanpy (e.g. is there some point where we take count data to the power of something and save it)
